app/models/concerns/programs_validator.rb

```diff
@@ -20,6 +22,13 @@ def validate_each(record, attribute, value)
 
   private
 
+  def check_hash(record, attribute, value)
+    return true if value.is_a?(Hash)
+
+    record.errors.add attribute, "is a #{value.class.name}, not a hash"
+    false
+  end
+
```

Comment:
What is a Program? Does it represent a method run on a robot or something?

Reply:
It represents a PCR program, which is a property of a primer panel. Essentially each primer panel has slightly different settings that it should be run under, which includes a name, and a duration. This gets displayed to the user when they are being asked to set-up the PCR reaction.

I've noticed that some of the more recent primer panels don't specify program names. That smells a bit wrong to me, as the PCR machine *will* have a program. I suspect it's partly because the program selected is already covered by the SOP, but might be other issues at play. (I know not all processes use the same two-stage PCR that the original did)

Reply:
OK thanks, yes I vaguely remember coming across these early in the Heron Limber pipeline development. I think I remember a conversation with R&D where they didn't care about the 'duration' stuff etc in this case, but can't remember why. 
Where is the interface where it gets displayed to the user? Is it just one of the plate screens in Limber?

Reply:
Yep, some of the plate presenters show the information, and we had an interstitial page on plate creation at one point, but I think that might have disappeared over the course of various pipeline changes.